lib/Sema/CSSimplify.cpp

```diff
@@ -6694,6 +6697,22 @@ static ConstraintFix *maybeWarnAboutExtraneousCast(
   // "from" could be less optional than "to" e.g. `0 as Any?`, so
   // we need to store the difference as a signed integer.
   int extraOptionals = fromOptionals.size() - toOptionals.size();
+
+  // From expression could be a type variable with a value to optional
+  // restrictions that we have to account for optionality mismatch.
+  const auto subExprType = cs.getType(castExpr->getSubExpr());
+  if (llvm::any_of(constraintRestrictions, [&](auto &entry) {
```

Comment:
Can we use `llvm::is_contained` here and simply pass a `std::make_tuple(type1, type2, ConversionRestrictionKind::ValueToOptional)`?

Reply:
Yeah, just tested and don't compile because `Type` doesn't have `==` operator :(

Reply:
We should have stored `TypeBase *` in that map actually because all we want is pointer equality...

Reply:
One option we could do to abstract that, is maybe have a `containsConversionRestriction(Type,Type, Kind)` in `ConstraintSystem`, not sure if it is worth it though... WDYT?

Reply:
Or switch `ConstraintRestrictions` to use `TypeBase *` :)

Reply:
Just adjusted @xedin =]